Zcash: add setConsensusBranchId() to TansactionBuilder

matthewzipkin-bitgo opened the pull request:
- adds `setConsensusBranchId` as property to `Transaction`
- sets that value by default based on transaction version
- updates the default values for ["Canopy" hard fork](https://zips.z.cash/zip-0251)
- use `setConsensusBranchId()` to maintain test success with old transaction version (from ["Heartwood" fork](https://zips.z.cash/zip-0250))
  - also see: https://github.com/BitGo/bitgo-utxo-lib/pull/78

TODO: 
- [x] Prepare consumers of this library to `setConsensusBranchId()` with Heartwood value until the day of the hard fork
- [x] Add test vectors for Canopy transactions

TICKET BG-24753

matthewzipkin-bitgo committed:
Update ZEC consensus-branch-id for "Canopy" hard fork

TICKET: BG-24753

matthewzipkin-bitgo committed:
Add consensusBranchId property to Transaction & set based on version

TICKET BG-24753

matthewzipkin-bitgo committed:
Add setConsensusBranchId() method to TansactionBuilder for Zcash

TICKET BG-24753

daira commented on src/transaction_builder.js:
```diff
@@ -524,10 +524,19 @@ TransactionBuilder.prototype.setVersion = function (version, overwinter = true)
+TransactionBuilder.prototype.setConsensusBranchId = function (consensusBranchId) {
+  if (!coins.isZcash(this.network)) {
+    throw new Error('consensusBranchId can only be set for Zcash transactions')
+  }
```
```suggestion
  }
  if (!this.inputs.every(function (input) { return input.signatures === undefined })) {
    throw new Error('No, this would invalidate signatures')
  }
```

matthewzipkin-bitgo replied:
Excellent thank you! updated in force-push to ed840b829079827c8d19458b4464e3a8ee498e90

matthewzipkin-bitgo replied:
I'm actually thinking of pushing back on this and removing the check. There's a few places downstream where this library is used to sign multisig in two stages. Both stages call `setConsensusBranchId()` to override the default, and the second one is failing because there's already one signature in place. If a caller does screw this up, they'll find out when they try to broadcast that one sig had the wrong `consensusBranchId`. I suppose a super advanced method would actually check the incoming function argument against any existing signatures to make sure we aren't mixing values up. This is a little different than changing the `locktime` for example if a tx already has a sig. In this case we aren't really invalidating the original sig, we're just possibly creating a transaction that won't pass consensus checks.

matthewzipkin-bitgo committed:
Add new test vector for tZEC Canopy P2SH 2-of-3

Also including README in test directory to explain how
to generate test vectors for the NEXT upgrade

TICKET BG-24753

matthewzipkin-bitgo commented:
@OttoAllmendinger @argjv this is ready for review and release -- with at least a minor version bump so consuming applications don't accidentally use the new value before the hard fork activates on mainnet. I'll work on updating those modules now.

Xekyo commented on test/fixtures/forks/zcash/README.md:
```diff
@@ -0,0 +1,126 @@
+# Generate test vectors for new ZCash testnet consensus branch IDs
```
Writing the procedure out here is great! :+1: 

Nit regarding the commit message of da9d22b0 (Add new test vector for tZEC Canopy P2SH 2-of-3), the "Ticket: BG-24753" must be the last line. IIRC, some Jira automation relies on the order here.

matthewzipkin-bitgo replied:
ah thanks, updated in 7434dc4b5b25c7939f409ba8f64c079f3d862aa4

matthewzipkin-bitgo commented on src/transaction_builder.js:
```diff
@@ -529,6 +529,17 @@ TransactionBuilder.prototype.setVersion = function (version, overwinter = true)
+    throw new Error('consensusBranchId can only be set for Zcash transactions')
+  }
+  if (!this.inputs.every(function (input) { return input.signatures === undefined })) {
+    throw new Error('No, this would invalidate signatures')
```
Thanks! updated and force pushed to 8a9c806ea75948843bf336d6224dd772a9ffef37

matthewzipkin-bitgo commented:
Moving this to https://github.com/BitGo/bitgo-utxo-lib/pull/90 to rebase on v1.7.1

OttoAllmendinger commented:
reopening this in order to fix `master` branch

reset this to 7434dc4 (before 1.7.x rebase), added my signatures and pushed again

OttoAllmendinger commented:
obsoleted by https://github.com/BitGo/BitGoJS/tree/master/modules/utxo-lib
